### PR TITLE
Initial work on updating status bar

### PR DESCRIPTION
**Pull request by jni**
# Description

This PR makes the following changes:

- active layer is now the topmost visible layer, regardless of selection. (See below for discussion.)
- printing of status messages based on value is now handled centrally, and much improved. I haven't changed over the obvious (integer) prints. In a future PR, I think this machinery can be used to print shape info, rather than just the shape index.
- if the cursor is out of range of the data, no value is shown, rather than showing the value at the nearest edge.
- a new keyboard shortcut is added, toggling visibility of the topmost layer.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How has this been tested?
- [x] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

# Discussion points

I'm still not sure about the active layer behavior. Now, it's a bit more cumbersome to mess with a layer's settings: for example, if I have a points layer and an image layer underneath it, I have to hide the points layer before I can adjust the clims on the image. However, this does solve the problem of adjusting settings on an invisible layer.

My suggestion, which I can implement here or in a followup PR depending on preferences, is that the active layer be defined as:
- if no layers are selected, the topmost visible layer.
- if a layer is selected, it is the active layer *if it is at all visible*, which means that only the following layers are in front of it:
  * image, labels, and pyramid with opacity < 1
  * points, shapes, or vectors, with any opacity (since these are generally sparse, visually)

I'll wait for thoughts before working on that logic.

Regarding the keyboard shortcut, topmost is kinda clunky, it turns out. My suggested improvement, which I can implement here or in a followup, is:

- Ctrl-V: hide the next visible layer (ie the active layer)
- Ctrl-Shift-V: show the next not-visible layer (ie the layer above the active layer), if any
- V: undo the last visibility change

Thoughts? And I just realised the shortcuts will have to change because Ctrl-V is taken. Suggestions welcome!

### CODE REVIEW

**Comment by jni** [comment]
Also, after this PR is resolved, I do want the status bar to show the status from *all* visible layers in a sequence. But that can come in a followup, and I could definitely use some guidance from @sofroniewn about how to go about this!

File: napari/layers/pyramid/pyramid.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
 import numpy as np
+
+from napari.util.status_messages import status_format
```
Comment (sofroniewn) [inline napari/layers/pyramid/pyramid.py]: relative path?

File: napari/layers/labels/labels.py [hunk 2]
```diff
@@ -495,7 +495,7 @@ def get_message(self, coord, value):
             String containing a message that can be used as a status update.
         """
         int_coord = np.round(coord).astype(int)
-        msg = f'{int_coord}, {self.name}, label {value}'
+        msg = f'{self.name} {int_coord} label {value}'
```
Comment (sofroniewn) [inline napari/layers/labels/labels.py]: do we need special handling for `None`?

File: napari/viewer.py [hunk 3]
```diff
@@ -18,3 +18,10 @@ def __init__(self, title='napari'):
         self.window = Window(qt_viewer)
         self.screenshot = self.window.qt_viewer.screenshot
         self.camera = self.window.qt_viewer.view.camera
+
+
+@Viewer.bind_key('v')
```
Comment (sofroniewn) [inline napari/viewer.py:24]: i feel like this should go in the `qt_viewer.py` file, not here - right @kne42?

**Review by sofroniewn — COMMENTED** [review]
I really like that you've consolidated the `status_format` into one function. Do you think we could go one step further and move the `get_message` function to the utils too? i.e. something like the following:

```python
def get_message(name, coord, values=None):
     # concatenate string name + convert coord to ints + concatenate values, where
     # if values is None then ignore, or if a dictionary then use string keys
     # and pass values through formatter, so could have
     # {'value' : (230, 221, 10)} or {'shape' : 10, 'vertex' : 2} for image / shapes
     # shapes layer respectively all turn out the way we want.
```

This could really cut back on duplicated code and make our layer model files much simpler.

I'll give the comments around visible / selected etc. more thought but my initial reaction is to like your suggestion around treating image-like and shape-like layers differently as there are inherent differences in their sparsity

File: napari/layers/pyramid/pyramid.py (hunk 1 above)
Comment (jni) [reply napari/layers/pyramid/pyramid.py]: Ah yes, good point, PyCharm does these imports automatically and I forgot to update this one.

Drives me nuts that they can't do relative imports!!! I might write a support ticket.

File: napari/layers/labels/labels.py (hunk 2 above)
Comment (jni) [reply napari/layers/labels/labels.py]: ... Yes, yes we do.

File: napari/viewer.py (hunk 3 above)
Comment (jni) [reply napari/viewer.py:24]: The layer shortcuts are all on the models... But you might be right!

**Comment by jni** [comment]
This is still WIP, but @sofroniewn @kne42, note that when I move my `@Viewer.bind_key` to `@ViewerModel.bind_key`, I can't get the keybindings to trigger at all... Is this expected?

**Comment by jni** [comment]
Also, @sofroniewn, I believe you were going to give me some pointers about where to start for concatenating status messages from multiple layers together? Apologies if you already did and I have lost track of it.

**Comment by sofroniewn** [comment]
@jni i would do the concatenation of status messages inside this for loop - https://github.com/napari/napari/blob/ebc5474a59a89496c99c43129b65e63afa7cc662/napari/components/viewer_model.py#L705

and then do `self.status = concat_status` at the end instead of `self.status = active_layer.status` and that will trigger all the updates needed

**Comment by sofroniewn** [comment]
@jni with the merging of #485 which incorporates some of the code from this PR I recommend we close this PR. Also I added similar keybindings from this PR in #472.

You can then make a new PR that makes all the status bar printing from many layers happen at the same time.

**Comment by jni** [comment]
Yep, that was the impression I had. Will deal with unifying status bars mid-september.

**Comment by sofroniewn** [comment]
Sounds good. I bumped unifying status bars to `0.2` so it's on our list in #301 now :-)

**Pull request closed without merging**